axiom/client.go

```diff
@@ -246,44 +265,64 @@ func (c *Client) newRequest(ctx context.Context, method, path string, body inter
 
 	// Set Content-Type.
 	if body != nil && !isReader {
-		req.Header.Set("Content-Type", "application/json")
+		req.Header.Set(headerContentType, mediaTypeJSON)
 	} else if body != nil {
-		req.Header.Set("Content-Type", "application/octet-stream")
+		req.Header.Set(headerContentType, defaultMediaType)
 	}
 
 	// Set authorization header, if present.
 	if c.accessToken != "" {
-		req.Header.Set("Authorization", "Bearer "+c.accessToken)
+		req.Header.Set(headerAuthorization, "Bearer "+c.accessToken)
 	}
 
 	// Set organization ID header when using a personal token.
 	if IsPersonalToken(c.accessToken) && c.orgID != "" {
-		req.Header.Set("X-Axiom-Org-Id", c.orgID)
+		req.Header.Set(headerOrganizationID, c.orgID)
 	}
 
 	// Set other headers.
-	req.Header.Set("Accept", "application/json")
-	req.Header.Set("User-Agent", c.userAgent)
+	req.Header.Set(headerAccept, mediaTypeJSON)
+	req.Header.Set(headerUserAgent, c.userAgent)
 
 	return req, nil
 }
 
 // do sends an API request and returns the API response. The response body is
 // JSON decoded or directly written to v, depending on v being an io.Writer or
 // not.
+// If the rate limit is exceeded and reset time is in the future, it returns
+// `*LimitError` immediately without making an API call.
 func (c *Client) do(req *http.Request, v interface{}) (*response, error) {
+	// If we've hit the rate limit, don't make further requests before
+	// the reset time.
+	if err := c.checkLimit(req); err != nil {
+		return &response{
+			Response: err.response,
+
+			Limit: err.Limit,
+		}, err
+	}
+
 	httpResp, err := c.httpClient.Do(req)
 	if err != nil {
 		return nil, err
 	}
-	defer httpResp.Body.Close()
+	defer func() {
+		_, _ = io.Copy(io.Discard, httpResp.Body)
+		_ = httpResp.Body.Close()
+	}()
 
-	resp := &response{httpResp}
+	resp := newResponse(httpResp)
+
+	key := limitKey(resp.Limit.limitType, resp.Limit.Scope)
+	c.limitsMu.Lock()
+	c.limits[key] = resp.Limit
+	c.limitsMu.Unlock()
```

Comment:
What if you get hold of the lock after another request updated a limit lower than what we have? Do we need a check here?

Reply:
You mean that request one is about to aquire the lock (and set the limit to `0`, meaning nothing remaining) while the concurrently running request two has already made it past `checkLimit()` and/or `Do()`?

In both of these cases, the limit is updated twice. However, I prefer that over dealing with additional locking as this is very unlikely to happen and does no harm. Worst case is sending a request even if a 429 has already been received.

Reply:
Well one way to do this would be wrapping line 319 in an `if resp.Limit > c.limits[key]`, right?

Reply:
Like this?

```suggestion
	key := limitKey(resp.Limit.limitType, resp.Limit.Scope)
	c.limitsMu.Lock()
	if resp.Limit.Limit > c.limits[key].Limit {
		c.limits[key] = resp.Limit
	}
	c.limitsMu.Unlock()
```

I fail to see how this prevents the described problem? `resp.Limit.Limit > c.limits[key].Limit` will be `true` all the time, no?